### DISCUSSION TITLE
Kobo Sync improvements

### DISCUSSION THREAD

**Discussion by Bart274** [gotson/komga#1677] [Ideas]
I have several possible improvements for the Kobo Sync feature:

- ~~Create a hash for each book based on it's metadata and use that hash as identifier for the Kobo Sync. By doing this, when you update the metadata of a book, the hash will change and when you sync your Kobo, it should no longer find the old hash and remove the book from the device and upload the new hash to the device with the new metadata. Only downside about this approach: if you have read that book partially on the Kobo in between syncs, you will lose that reading progress.~~
- Allow users to select which libraries are synced to their Kobo device.
- Allow users to select per library if they want all books to be synced or only the unread books (for example, I want all my books to be visible on my Kobo, but only my unread comics)

What could improve the Kobo Sync feature as well is a big change for Komga: allow multiple files for the same book/comic like Plex allows multiple files to be matched for one movie.
It would allow us to have cbz files for comics which can be used on all comic readers but have an epub version as well of the same comic for Kobo sync (and other epub readers).
The reason why I would love this is that I'm reading my cbz files with Panels, but I would also want to read them on my Kobo. Now I have 2 separate libraries for this, a cbz library for Panels and a epub library for Kobo. But these items are unrelated to eachother, so if I have read a comic in Panels, it's marked as read in Komga, but only the cbz version, the epub version remains unread.

**Comment by gotson** [comment]
> * Create a hash for each book based on it's metadata and use that hash as identifier for the Kobo Sync. By doing this, when you update the metadata of a book, the hash will change and when you sync your Kobo, it should no longer find the old hash and remove the book from the device and upload the new hash to the device with the new metadata. Only downside about this approach: if you have read that book partially on the Kobo in between syncs, you will lose that reading progress.

i tried this already, this is fairly simple to do but doesn't work in practice. The Kobo will indeed update metadata, but lose read progress, and will have to download cover and file again.

I don't see how this could be used practically, as it cannot be turned on by default.

> Allow users to select which libraries are synced to their Kobo device.

this is planned, but tied up to other enhancements unrelated to Kobo, and might take some time to get there.



> Allow users to select per library if they want all books to be synced or only the unread books (for example, I want all my books to be visible on my Kobo, but only my unread comics)

Can you explain the rationale behind this ?

> What could improve the Kobo Sync feature as well is a big change for Komga: allow multiple files for the same book/comic like Plex allows multiple files to be matched for one movie.
> It would allow us to have cbz files for comics which can be used on all comic readers but have an epub version as well of the same comic for Kobo sync (and other epub readers).

This was already discussed in https://github.com/gotson/komga/issues/497 but there's no plans to support that. I believe Komga should be able to convert on the fly depending on the destination, instead of asking the user to manage the different pre-converted files upstream.

**Reply by Bart274** [reply]
> > * Create a hash for each book based on it's metadata and use that hash as identifier for the Kobo Sync. By doing this, when you update the metadata of a book, the hash will change and when you sync your Kobo, it should no longer find the old hash and remove the book from the device and upload the new hash to the device with the new metadata. Only downside about this approach: if you have read that book partially on the Kobo in between syncs, you will lose that reading progress.
> 
> i tried this already, this is fairly simple to do but doesn't work in practice. The Kobo will indeed update metadata, but lose read progress, and will have to download cover and file again.
> 
> I don't see how this could be used practically, as it cannot be turned on by default.

The question is also: how often do people change the metadata of a book/comic. It could be an option to have a checkbox on each item 'Metadata complete' and only sync books with this flag active.

> > Allow users to select per library if they want all books to be synced or only the unread books (for example, I want all my books to be visible on my Kobo, but only my unread comics)
> 
> Can you explain the rationale behind this ?

I sometimes want to reread a book that I already read in the past. If the book is still on the Kobo, it's easier to start reading it again.
Comics I usually read one series at a time and once they are read, I don't want to see them on my device again until I marked the entire series as unread again, the read status is used to keep track which item I should read next.
Also I want my stats on my kobo to reflect my book reading times, having the comics on the Kobo after reading them will have those stats 'mess up' my book reading stats.

> > What could improve the Kobo Sync feature as well is a big change for Komga: allow multiple files for the same book/comic like Plex allows multiple files to be matched for one movie.
> > It would allow us to have cbz files for comics which can be used on all comic readers but have an epub version as well of the same comic for Kobo sync (and other epub readers).
> 
> This was already discussed in #497 but there's no plans to support that. I believe Komga should be able to convert on the fly depending on the destination, instead of asking the user to manage the different pre-converted files upstream.

On the fly conversion seems like a good idea, but then you need to support multiple conversion tools. For example kepubify works fine when I want to convert an epub to kepub, but it messes up the images in cbz files. For comics it's better to use a tool like KCC, but there are a lot of different profiles, for each device type a different profile exists and you would need to use grayscale or colored conversion.
I think it would require the system users to set the path for kepubify and kcc, in your on-the-fly conversion you would need to use kepubify for epub files and kcc for comic files (but kcc is slow). And it would require people to be able to set the right profile and other options for the KCC conversion in their API key settings (so you could use multiple kobo devices, but each has it's own api key)

**Reply by gotson** [reply]
> The question is also: how often do people change the metadata of a book/comic. It could be an option to have a checkbox on each item 'Metadata complete' and only sync books with this flag active.

this is way too complicated imho. Kobo should fix their shit :D

> Also I want my stats on my kobo to reflect my book reading times, having the comics on the Kobo after reading them will have those stats 'mess up' my book reading stats.

can you expand on that ? i am not familiar with kobo reading stats, or why having books on the kobo would mess that up.

> On the fly conversion seems like a good idea, but then you need to support multiple conversion tools. For example kepubify works fine when I want to convert an epub to kepub, but it messes up the images in cbz files. For comics it's better to use a tool like KCC, but there are a lot of different profiles, for each device type a different profile exists and you would need to use grayscale or colored conversion.
> I think it would require the system users to set the path for kepubify and kcc, in your on-the-fly conversion you would need to use kepubify for epub files and kcc for comic files (but kcc is slow). And it would require people to be able to set the right profile and other options for the KCC conversion in their API key settings (so you could use multiple kobo devices, but each has it's own api key)

that is already in the solution space, i have ideas already to implement this without any hassle for the user

**Reply by Bart274** [reply]
> > The question is also: how often do people change the metadata of a book/comic. It could be an option to have a checkbox on each item 'Metadata complete' and only sync books with this flag active.
> 
> this is way too complicated imho. Kobo should fix their shit :D

Ok. :-) Does the force sync functionality refreshes all metadata?

> > Also I want my stats on my kobo to reflect my book reading times, having the comics on the Kobo after reading them will have those stats 'mess up' my book reading stats.
> 
> can you expand on that ? i am not familiar with kobo reading stats, or why having books on the kobo would mess that up.

As you can see below, you have reading statistics on your kobo. It shows you the last opened book, the progress in that book and the progress in the current chapter.
At the bottom it shows you an overall view of your entire library, but this also taking into account all your comics that are read. The estimated reading times for books are based on your overall reading times, but I read a comic page a lot faster than a book page (as I suspect everyone does).
![kobo stats](https://github.com/user-attachments/assets/3d9164bb-ddb2-402a-9b5a-464cf13f96f1)

> > On the fly conversion seems like a good idea, but then you need to support multiple conversion tools. For example kepubify works fine when I want to convert an epub to kepub, but it messes up the images in cbz files. For comics it's better to use a tool like KCC, but there are a lot of different profiles, for each device type a different profile exists and you would need to use grayscale or colored conversion.
> > I think it would require the system users to set the path for kepubify and kcc, in your on-the-fly conversion you would need to use kepubify for epub files and kcc for comic files (but kcc is slow). And it would require people to be able to set the right profile and other options for the KCC conversion in their API key settings (so you could use multiple kobo devices, but each has it's own api key)
> 
> that is already in the solution space, i have ideas already to implement this without any hassle for the user

Ok, glad to hear that :-) Curious to see how it will work out :-)

**Reply by gotson** [reply]
> Ok. :-) Does the force sync functionality refreshes all metadata?

i don't think so. What it does is that it deletes the Komga history of what has been synced for this API key in the past. Meaning on the next sync, all of the books will be returned as NewEntitlements. I don't sync Kobo will update metadata if the book is there already.

> As you can see below, you have reading statistics on your kobo. It shows you the last opened book, the progress in that book and the progress in the current chapter.
> At the bottom it shows you an overall view of your entire library, but this also taking into account all your comics that are read. The estimated reading times for books are based on your overall reading times, but I read a comic page a lot faster than a book page (as I suspect everyone does).

that doesn't explain how having already read comics in the Kobo impacts the stats though

**Reply by Bart274** [reply]
> > Ok. :-) Does the force sync functionality refreshes all metadata?
> 
> i don't think so. What it does is that it deletes the Komga history of what has been synced for this API key in the past. Meaning on the next sync, all of the books will be returned as NewEntitlements. I don't sync Kobo will update metadata if the book is there already.

Ok thanks.

> > As you can see below, you have reading statistics on your kobo. It shows you the last opened book, the progress in that book and the progress in the current chapter.
> > At the bottom it shows you an overall view of your entire library, but this also taking into account all your comics that are read. The estimated reading times for books are based on your overall reading times, but I read a comic page a lot faster than a book page (as I suspect everyone does).
> 
> that doesn't explain how having already read comics in the Kobo impacts the stats though

Well, having read comics on the Kobo (I mean visible on the Kobo independent on whether or not it's downloaded), will make sure these read comics are taken into account for your reading speed, so your estimated time needed to finish a book is impacted by these comics. But I think it's just my problem. :-)
If you add more control options for people on what they want to sync, for example in a specific library I only want to sync specific reading lists, my problem will be solved as well as I have reading lists for my comics with only the unread issues.

**Reply by luukrijnbende** [reply]
> Allow users to select which libraries are synced to their Kobo device.

> this is planned, but tied up to other enhancements unrelated to Kobo, and might take some time to get there.

Any idea on the timeline of this? I plan on sharing the library with multiple people that all can request their own books, so I wouldn't like to have all of the book in the library 'available' on the Kobo.

**Comment by Bart274** [comment]
Would it be possible to set the title of books/comics in series for Kobo to be: `<series> <zero padded issue number> <title>`?
This way all books in a series would be ordered neatly on the Kobo.

**Reply by gotson** [reply]
> Would it be possible to set the title of books/comics in series for Kobo to be: `<series> <zero padded issue number> <title>`? This way all books in a series would be ordered neatly on the Kobo.

the kobo already has a series view you can use

**Reply by Bart274** [reply]
I know there is a Series view, but I wanted to organize the Books view. But damnit, I'm stupid. I forgot to switch the sorting from title to series :-D That solves my issue completely so feel free to ignore my stupidity :-D

**Comment by PhilippKammann** [comment]
- **Support for Collections. Especially a "On Deck" collection.**

My suggestion would be for Komga to create collections automatically and sync them to Kobo

**Reply by Bart274** [reply]
I'm not a fan of automatically created reading lists or collections. But the ability to limit what to be synced would be sweet indeed.
The "On Deck" collection is the only thing I wouldn't mind, but then it has to be configurable, so you could for example have the first 5/10/... items from a series on the "On Deck", so the next items are there to be able to download them in advance before you go offline.

**Reply by gotson** [reply]
> My suggestion would be for Komga to create collections automatically and sync them to Kobo

what do you mean by creating collections automatically ? Komga already has collections/readlists.

Can you maybe clarify how that would work with the Kobo ?

**Reply by PhilippKammann** [reply]
I mean Kobo-Collections. These are user-generated groupings of books on the Kobo device (by, e.g., genre or themes). Komga's equivalent would be an unordered readlist.

On Kobo, you would manually create a Kobo-Collection and add books one by one.

My feature request would be to have Komga create Kobo-Collections and sync them to the Kobo.

Calibre-Web does this and syncs CW's "Shelf"s as Kobo-Collections

**Reply by gotson** [reply]
CW's Shelves are akin to Komga's readlists.

> My feature request would be to have Komga create Kobo-Collections and sync them to the Kobo.

If Komga could create Kobo collection for readlists that exists in Komga and are part of the sync already, would that work ? That was my initial idea.

**Reply by PhilippKammann** [reply]
Yeah. That would work.

I would also like to have "On Deck" on the Kobo, but then I can just make a automatic readlist with Komga's Api

**Reply by gotson** [reply]
> I would also like to have "On Deck" on the Kobo, but then I can just make a automatic readlist with Komga's Api

i think that makes sense, we could have on deck with the books that are part of the sync

**Reply by gotson** [reply]
On Deck has been added in the latest release

**Comment by SenorSmartyPants** [comment]
- Preserve annotations

Thank you so much for implementing Kobo Sync. I tried using Calibre-web's Kobo Sync feature and I ran into this issue

https://github.com/janeczku/calibre-web/issues/2610

My Kobo user likes to make notes in the books they read, so I can't set up sync since the notes get deleted. Do you have any plans to look into this issue?

Thanks again for making Komga in the first place.

**Reply by gotson** [reply]
I would need to look into it,i never used them so i might need to try it first.

**Reply by gotson** [reply]
I checked on this today, first off, you need KEPUB files for the kobo to even store the notes. Komga does not convert EPUB to KEPUB (yet), thus you would need to have KEPUB files to start with (with an `.epub` extension so Komga can scan them).

From my tests, the annotations are not synced using Kobo Sync, at least Komga is not receiving any API call about annotations. When doing a sync, even if the book was changed on Komga side, the annotations remain on the Kobo.

**Reply by geneccx** [reply]
Check out the issue here: https://github.com/janeczku/calibre-web/issues/2155

The reason komga doesn't receive an API call is because it's set under `reading_services_host` under the initialization endpoint, which is defaulted to `https://readingservices.kobo.com`. If you override it to Komga's host it should then come through.

**Reply by gotson** [reply]
as i explained in #1692 i cannot reproduce this behavior. Annotations do not disappear on my end. I tried with a Nia and a Libra Colour.

**Reply by geneccx** [reply]
Yeah, that I'm not sure about either. The reading services endpoint from Kobo allows for arbitrary book ids and annotations, so it should theoretically work. Just mentioning that was why you're not seeing the Kobo Sync requests.

**Reply by gotson** [reply]
thanks, for clarification there are 2 topics:
1. on calibre web, annotations are not working because the `reading_services_host` is not present at all if kobo proxying is disabled. This does not happen on Komga, because the default response from the initialization call contains this key with the kobo reading services URL.
2. I understand there is another discussion altogether on kobo patches repo about using a different service altogether, but for the moment this is not something i want to pursue for Komga, since the Kobo service does it's job already it seems.

**Comment by Mattiams** [comment]
I have some ideas to share regarding the Kobo Sync feature. I've spoken briefly with Patji. on the Discord Server, who suggested I post here. I'm aware of some of what I am about to suggest being brought up before in this [Feature Request](https://github.com/gotson/komga/issues/371):

- Kobo Sync only works with the EPUB/KEPUB format. CBR/CBZ will not work. When converting a Comic to EPUB/KEPUB format, either via [Kindle Comic Converter](https://github.com/ciromattia/kcc) or [Kombo](https://github.com/soda3x/Kombo), two issues occur:
1.  The file will show twice in the Komga UI.
2. Metadata applied via Komf for example (or other tools) will leave areas blank. As well as leaving a 'Null' red textbox in the comment area. 

My question is:
- Can a comic issue show multiple file formats? For instance offering up a EPUB/KEPUB for Kobo Sync, whilst retaining a CBR/CBZ instance for OPDS applications on other devices?
- Or, can a CBR/CBZ be converted on the fly to the EPUB/KEPUB format for Kobo Sync?

I've included screenshots to illtustrate what I'm rambled about above.
Here is the link to the [original Discord post](https://discord.com/channels/678794935368941569/678795177393127424/1370857925165449287) I made.
![CBZ Instance](https://github.com/user-attachments/assets/fed5c7f9-4109-4843-94f4-c1e16c30da0a)
![EPUB Instance](https://github.com/user-attachments/assets/9c770346-b598-41d0-bf00-99d8a12177c7)

**Reply by gotson** [reply]
on the fly conversion is planned
